Render developer replies to reviews

kumar303 opened the pull request:
Fixes https://github.com/mozilla/addons-frontend/issues/2773

* In order to keep the patch small I *did not* update the form UI. That will happen in https://github.com/mozilla/addons-frontend/issues/3370
* I didn't adjust the review listing layout beyond what was necessary to render the reply. The listing layout will be adjusted in ﻿﻿https://github.com/mozilla/addons-frontend/issues/2564

Before these changes, the developer reply was not rendered at all. Here are the after screenshots:


<img width="800" alt="screenshot 2017-10-06 11 26 33" src="https://user-images.githubusercontent.com/55398/31288345-29b0e2ca-aa8a-11e7-8208-284007c35a3f.png">
<img width="799" alt="screenshot 2017-10-06 11 28 26" src="https://user-images.githubusercontent.com/55398/31288350-2cf76256-aa8a-11e7-8b25-8e6ade56f759.png">
<img width="320" alt="screenshot 2017-10-06 11 26 06" src="https://user-images.githubusercontent.com/55398/31288351-2f527ca2-aa8a-11e7-8cb5-9d4c3f60226d.png">

tofumatt commented on src/ui/components/Icon/styles.scss:
```diff
@@ -164,3 +164,15 @@
+  width: 14px;
+
+  [dir=rtl] & {
+    // Flip the horizontal orientation of the arrow.
```
👍

tofumatt commented on tests/unit/amo/components/TestAddonReviewListItem.js:
```diff
@@ -442,4 +460,88 @@ describe(__filename, () => {
+      expect(root.find(Rating)).toHaveLength(0);
+    });
+
+    it('does not include a user name in the by line', () => {
```
I thought byline was one word?

tofumatt commented on tests/unit/amo/components/TestAddonReviewListItem.js:
```diff
@@ -442,4 +460,88 @@ describe(__filename, () => {
+        .not.toIncludeText(reply.userName);
+    });
+
+    it('hides the edit button for replies', () => {
```
Can developers not edit their replies? Huh. Seems like it would be useful.

tofumatt replied:
Oh I see, waiting on https://github.com/mozilla/addons-frontend/issues/3368. A TODO here about it might be handy.

kumar303 replied:
The TODO with a link to the issue in in the code. I think that's a better place to put TODOs. When you search for the TODO and remove the code, the test will fail so you will know which tests to delete. However, if you search for a TODO that leads to a test and you delete the test, you might not know which code needs updating.

willdurand commented:
I really like the UI 😍

Maybe the `margin-start` is not necessary on small screens though because we already have a clear separation of the different reviews ans it wastes useful space.

I also wonder if the arrow should not be horizontally inverted for "Reply from the developer". It is never clear to me what this curved arrows mean to be honest, but here I would expect them to point to the reviews that have been replied to.

kumar303 commented on src/amo/components/AddonReviewListItem/styles.scss:
```diff
@@ -1,41 +1,89 @@
+  border-radius: 12px;
+  margin-left: 10px;
+  margin-top: 23px;
+  padding: 19px 23px 17px;
```
I found out how to measure pixel distances in Sketch! http://www.sketchtips.info/articles/mind-your-shortcuts

kumar303 replied:
On second thought, I'm pretty sure I just measured wrong, heh. I flipped everything back to the standard 6px, 12px, and 24px specs. I think it looks better.

kumar303 commented on src/amo/components/AddonReviewListItem/styles.scss:
```diff
@@ -1,41 +1,89 @@
   a:hover,
   a:active {
-    color: $link-color;
+    color: $review-reply-color;
```
ah, thanks. I was searching for the hex which is why I couldn't find it.

kumar303 commented on src/amo/components/AddonReviewListItem/styles.scss:
```diff
@@ -1,41 +1,89 @@
+
+.AddonReviewListItem-control {
+  display: flex;
+  margin-left: 5px;
```
I always wondered what those helpers did!

kumar303 commented on src/amo/components/AddonReviewListItem/styles.scss:
```diff
@@ -1,41 +1,89 @@
+}
+
+.AddonReviewListItem-reply {
+  background-color: transparentize($review-reply-color, 0.95);
```
In Sketch it's the `$blue-50` hex value set to 5% opacity. Is it better to add a custom hex value for it?

tofumatt replied:
Ah, I guess keep with the transparent one actually, it's more flexible. 👍

kumar303 commented:
Thanks for the review. I changed all spacing to match the standard 6px, 12px, and 24px specs. I updated the screenshots.

> Maybe the margin-start is not necessary on small screens

Good call. Fixed.

> I also wonder if the arrow should not be horizontally inverted for "Reply from the developer". 

To me it's using the reader's natural reading direction (left-to-right, for example) to indicate this is pointing back to the original source. Thus, I think it would make sense to reverse the direction for right-to-left readers.
